### ISSUE TITLE
Smooth transition in BoxTrackerCalculator never worked?

### ISSUE THREAD

**Issue by shrubb** [google/mediapipe#1238]
Hello MediaPipe team, and thanks for the brilliant framework!

So I'm running the `objecttrackinggpu` Android example. When each new detection arrives, boxes "jump" rapidly, which I'm trying to alleviate.

Fortunately, there's a dedicated `start_pos_transition_frames` field in `BoxTrackerCalculator` just for that. However, it gives no effect. I spent 3-4 hours debugging [the C++ code](https://github.com/google/mediapipe/blob/a8d6ce95c4945087dc060760bad6258a7876c76e/mediapipe/calculators/video/box_tracker_calculator.cc#L1209) only to learn that

1. "fast forwarding" never happens because the detections' timestamps are larger than those of all previous frames (fair enough);
2. because of that (**which to me seems wrong**), `start_pos_list` is always empty.

Either it's a bug or I'm not using `start_pos_transition_frames` properly :thinking:

**Comment by lmzh123** [comment]
Hey @shrubb,

I've also been dealing with the box initialization and tracking data being out of sync issue, which I believe is what you call "Jump" when entering into the **BoxTrackerCalculator**. I was not aware of the `start_pos_transition_frames` option, but I agree with you that it seems to do nothing and for now the only way I get the synchronization among them is to move the `START_POS` signal to the same `sync_set` of the other signals, which makes the tracking feed to go to detections feed rate.

```
input_stream_handler {
  input_stream_handler: "SyncSetInputStreamHandler"
  options {
    [mediapipe.SyncSetInputStreamHandlerOptions.ext] {
      sync_set {
        tag_index: "TRACKING"
        tag_index: "TRACK_TIME"
        tag_index: "START_POS"
      }
      sync_set {
        tag_index: "CANCEL_OBJECT_ID"
      }
    }
  }
}
```

Let me know if you find a way to solve this issue.

**Comment by shrubb** [comment]
@lmzh123, thanks for your advice! I tried it, unfortunately nothing changed, still getting the "Received start pos beyond current timestamp" warning.

I'll dig deeper and keep posted. Possibly I'll just make a dirty workaround in C++.

**Comment by shrubb** [comment]
Investigation complete, finally.

### TL;DR

* MediaPipe's built-in box handling needs strict conventions and a simpler structure.
* I had to tweak some default calculators to make them work in scenarios they were probably designed for.
* In addition to that, to properly prevent boxes from "leaping" with new detections in `objecttrackinggpu`, there ideally has to be a dedicated detection matching calculator in the graph.

### Details for Nerds

It took me quite a while to understand all the classes and how they work together. The MediaPipe code turned out to be **very** general with lots of abstractions.

Box tracking isn't an exception, likely because it anticipates possible future applications (e.g. loop closure in SLAM). I found at least **8** classes that represent bounding boxes in various contexts!

Namely, there are 2 important classes used in `TrackedDetectionManagerCalculator`:

* `TimedBoxProto`: (ideally) represents a unique *object* in an image, and has a field `id` meaning *object ID* (*"the person #⁠2"*);
* `TrackedDetection`: (ideally) represents a unique *bounding box*, and has a field `unique_id` which is the unique number referring to a *bounding box*, as well as `previous_id` which is in fact the `unique_id` of this object's previous box (*"the object detector has output the box #⁠1337, which is known to bound the same object that was until now bounded by the box #⁠1234"*).

These make perfect sense. The problem? **`TrackedDetectionManagerCalculator` doesn't adhere to the above convention**. It mixes up these two classes and uses them interchangeably, effectively treating all detections as different objects (whose IDs `0, 1, 2, 3...` in our case come from the `DetectionUniqueIdCalculator`):

https://github.com/google/mediapipe/blob/f15da632dec186f2c1d3c780f47086477e2286a9/mediapipe/calculators/video/tracked_detection_manager_calculator.cc#L245-L252

https://github.com/google/mediapipe/blob/f15da632dec186f2c1d3c780f47086477e2286a9/mediapipe/calculators/video/tracked_detection_manager_calculator.cc#L268-L276

https://github.com/google/mediapipe/blob/f15da632dec186f2c1d3c780f47086477e2286a9/mediapipe/calculators/video/tracked_detection_manager_calculator.cc#L219

There's no quick global fix to this. One will at least have to e.g. retrofit those bounding box classes with the ID of the object they relate to (and/or vice versa), and change some default calculators accordingly.

However, here are two dirty fixes I tried in context of `objecttrackinggpu`:

1. **Tweak calculators' interfaces**. When `TrackedDetectionManagerCalculator` discovers a duplicate, an obsolete, or an out-of-view box, it communicates its ID in a `CANCEL_OBJECT_ID` packet. This is misleading, because it's not an object ID but a box ID. Well, if so, here's a workaround. In `CANCEL_OBJECT_ID`, let's also ship the ID of the detection in favour of which the old one was canceled. Then in `BoxTrackerCalculator` for each new detection we'll finally know what was its `previous_id` — and this will automatically revive `AddSmoothTransitionToOutputBox`. [Here is my dirty implementation](https://github.com/shrubb/mediapipe/tree/start-pos-smoothing) of that trick; it's not perfect, boxes flicker because `CANCEL_OBJECT_ID` arrive before new detections, but you get the point.

1. **Limit the graph to one object**. Specifically, pretend that all detection IDs are object IDs. This is a more natural solution, but it only works when we have a known box-object correspondence. For instance, that correspondence can be established by a dedicated calculator that does object re-identification or just rough box matching. [I didn't bother](https://github.com/shrubb/mediapipe/tree/single-detection-hard-pick) and just replaced `DetectionUniqueIdCalculator` with my custom `SingleDetectionCalculator` that throws away all boxes but one, and sets that detection's object ID to `1`. `TrackedDetectionManagerCalculator` also needed a small tweak because it's only suited to unique box IDs.

### What Next?

At first I wanted to make a pull request, but right now that's just too much work for me. So maybe. At least let the gist be summarized here, hope it will help someone.

**Comment by sgowroji** [comment]
Hi @shrubb, Thank you so much for adding your observation. This will really help the community users who ever looking for similar queries. Presently we are closing this issue and still if any have other relevant queries in the above context, Feel free to reopen the issue. Thanks!
